### PR TITLE
Feat : ANF (Augmented Normalizing Flows)

### PR DESCRIPTION
**Pull request by Namgyu-Youn**
## Description
I builded ANF introduced in #28. (See also : [Arxiv - ANF](https://arxiv.org/abs/2002.07101))

## Testing module
I tested using `test_anf.py` (following code).
```python
import numpy as np
import torch

from zuko.flows import ANF, HierarchicalANF


def toy_example():
    # Set random seed for reproducibility
    torch.manual_seed(42)

    # Create a simple ANF model
    features, noise_features, context_size = 2, 2, 4  # 2D data, 2D noise, Optional context dimension

    # Create model with 3 autoencoding transform layers
    model = ANF(
        features=features,
        noise_features=noise_features,
        context=context_size,
        steps=3,
        hidden_features=[64, 64],  # Hidden layer sizes for the networks
    )

    # Generate dummy context
    context = torch.randn(context_size)

    # Sample from the model
    n_samples = 1000
    samples_with_noise = model(context).sample((n_samples,))

    # Extract only the data part (not the noise)
    samples = samples_with_noise[:, :features]

    print(f"Generated {n_samples} samples with shape: {samples.shape}")

    # Calculate log probability of the samples
    log_probs = []
    for i in range(5):  # Just check a few samples
        # Generate random noise for the sample
        noise = torch.randn(noise_features)
        sample_with_noise = torch.cat([samples[i], noise])

        # Calculate log probability
        log_prob = model(context).log_prob(sample_with_noise)
        log_probs.append(log_prob.item())

def hierarchical_example():
    """Example with hierarchical ANF."""
    print("\nRunning hierarchical ANF example...")

    # Set random seed for reproducibility
    torch.manual_seed(42)

    # Create a hierarchical ANF model
    features = 2  # 2D data
    noise_features_list = [2, 3, 2]  # Three levels of latent variables
    context_size = 4  # Optional context dimension

    # Create hierarchical model
    model = HierarchicalANF(
        features=features,
        noise_features_list=noise_features_list,
        context=context_size,
        hidden_features=[64, 64],  # Hidden layer sizes for the networks
    )

    print(f"Created hierarchical model with {len(noise_features_list)} levels")

    # Generate dummy context
    context = torch.randn(context_size)

    # Sample from the model
    n_samples = 1000
    samples_with_noise = model(context).sample((n_samples,))

    # Extract only the data part (not the noise)
    samples = samples_with_noise[:, :features]

    print(f"Generated {n_samples} samples with shape: {samples.shape}")

def density_estimation_example():
    """Demonstrate density estimation on a 2D toy dataset."""
    print("\nRunning density estimation example...")

    # Set random seed for reproducibility
    torch.manual_seed(42)

    # Create a toy 2D mixture of Gaussians dataset
    def sample_mog(n_samples=1000, n_components=5):
        """Generate samples from a mixture of 2D Gaussians."""
        # Fixed centers for the Gaussians
        centers = [
            [-4, -4],
            [-4, 4],
            [4, -4],
            [4, 4],
            [0, 0]
        ][:n_components]

        # Generate samples
        samples = []
        for _ in range(n_samples):
            # Choose a component
            k = np.random.randint(n_components)
            # Sample from that Gaussian
            x = np.random.normal(centers[k][0], 0.5)
            y = np.random.normal(centers[k][1], 0.5)
            samples.append([x, y])

        return torch.tensor(samples, dtype=torch.float32)

    # Generate dataset
    n_samples = 1000
    data = sample_mog(n_samples=n_samples)

    print(f"Generated {n_samples} samples for training")

    # Create ANF model for density estimation
    features, noise_features = 2, 2

    model = ANF(
        features=features,
        noise_features=noise_features,
        steps=5,
        hidden_features=[128, 128],
    )

    # Train the model (simplified training loop)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    n_epochs, batch_size = 100, 64

    print(f"Training ANF model for {n_epochs} epochs...")

    for epoch in range(n_epochs):
        # Shuffle data
        perm = torch.randperm(n_samples)
        running_loss = 0.0

        # Train in batches
        for i in range(0, n_samples, batch_size):
            indices = perm[i:i+batch_size]
            x_batch = data[indices]

            # Generate random noise
            e_batch = torch.randn(len(indices), noise_features)

            # Combine data and noise
            x_e_batch = torch.cat([x_batch, e_batch], dim=1)

            # Zero the gradients
            optimizer.zero_grad()

            # Compute loss (negative log-likelihood)
            log_prob = model().log_prob(x_e_batch)
            loss = -log_prob.mean()

            # Backpropagation
            loss.backward()
            optimizer.step()

            running_loss += loss.item() * len(indices)

        epoch_loss = running_loss / n_samples

        if (epoch + 1) % 20 == 0:
            print(f"Epoch {epoch+1}/{n_epochs}, Loss: {epoch_loss:.4f}")

    print("Training completed!")

    # Generate samples from the trained model
    with torch.no_grad():
        samples_with_noise = model().sample((n_samples,))

if __name__ == "__main__":
    toy_example()
    hierarchical_example()
    density_estimation_example()
```
And the result is : 
image

### CODE REVIEW

**Comment by francois-rozet** [comment]
Hi @Namgyu-Youn, thank you very much for the PR. Augmented flows is a long awaited feature! Unfortunately, I will not have time to review your PR until the end of March. Thank you again for your contribution!

**Comment by Namgyu-Youn** [comment]
Thanks for your comment. I would want to implement **Gaussianization Flows** ([Arxiv](https://arxiv.org/abs/2003.01941)) introduced in #28 until that time. Is there any advice or opinion??

p.s. Good luck to your career. I would gratefully wait for it!!

**Comment by francois-rozet** [comment]
Gaussianization Flow (GF) is already implemented in the library. 

P.S. Thanks !

**Comment by francois-rozet** [comment]
Hi @namgyu-youn, sorry for the wait.

I just checked the code, and I don't understand it at all. It is bad practice to create functions and classes dynamically. Could you explain why this is necessary in this case? More generally could you describe your implementation at a high level?

Also, Zuko components should be documented and tested.

**Comment by namgyu-youn** [comment]
> Hi @namgyu-youn, sorry for the wait.
> 
> I just checked the code, and I don't understand it at all. It is bad practice to create functions and classes dynamically. Could you explain why this is necessary in this case? More generally could you describe your implementation at a high level?

@francois-rozet The dynamic class was intended to handle the **complex conditional transformations**:
1. Encoding transformation: transforms noise e to z conditioned on data x (e|x → z)
2. Decoding transformation: transforms data x to y conditioned on transformed noise z (x|z → y)

Transform (`torch.distribution`) is designed for straight-forward, self-contained transformations. However, ANF requires handling  two **conditional transformations**. It was challenge to implement ANF using straight-forward methods, and I adopted encapsulating the complex, multi-step process for clarify it; As you mentioned, this class could be refactored using other methods, but I am not certain if I can handle it... 

> Also, Zuko components should be documented and tested.

If `documented` means more markdowns (e.g., See also, References, Arguments, Examples), then I will gratefully update it.

I will surely build pytest module, but I have learned huge PR (feature + bugfix + pytest + ...) was not helpful for some reasons (e.g., rebase, duplicate review, delay). So I want to carefully suggest seperating PR (feature - pytest)

**Comment by namgyu-youn** [comment]
Also, could you review for [b2b7765](https://github.com/probabilists/zuko/pull/61/commits/b2b7765721104109415f149b43f041c96c3f4f9d)? `.pre-commit-config.yaml` is more standard name for pre-commit, so renaming it and updating [CONTRIBUTING.md - Ln 30](https://github.com/probabilists/zuko/blob/master/CONTRIBUTING.md) like the following could be useful. : 
` pre-commit install --config pre-commit.yml` -> `pre-commit install`

**Comment by francois-rozet** [comment]
> Also, could you review for [b2b7765](https://github.com/probabilists/zuko/pull/61/commits/b2b7765721104109415f149b43f041c96c3f4f9d)?

I chose `pre-commit.yml` on purpose because I don't like this file to be hidden, while the other config files (`pyproject.toml`, `readthedocs.yml`, ...) or not. Using a non-standard name also makes the installation process more explicit. So, please revert this commit.

**Comment by francois-rozet** [comment]
Hi @namgyu-youn, I took a deeper look into your implementation. Unfortunately, it is way too convoluted and not up to the standards of Zuko. There is no need to implement ANF at the transformation level. An ANF is just a NF with additional (noise) features. It can therefore be a simple wrapper around a `Flow` instance that implements custom sampling and log_prob methods to handle the additional (noise) features automatically (the user should not need to sample noise by hand).

Note that in the paper, Fig 2a represents a sequence of two coupling transforms and Fig 2c represents a sequence of two auto-regressive transforms. These are already implemented in Zuko.

If you wish to contribute to Zuko, maybe you should start with a feature that is easier to add. For example, the rational-linear spline transformation (see https://arxiv.org/abs/2001.05168) should be easy to implement. You can take inspiration from the rational-quadratic spline transformation, already present in the library:

https://github.com/probabilists/zuko/blob/6fd97cbd5d3a394be01a90ce094cb2343383a7f5/zuko/transforms.py

**Comment by namgyu-youn** [comment]
> Hi @namgyu-youn, I took a deeper look into your implementation. Unfortunately, it is way too convoluted and not up to the standards of Zuko. There is no need to implement ANF at the transformation level. An ANF is just a NF with additional (noise) features. It can therefore be a simple wrapper around a `Flow` instance that implements custom sampling and log_prob methods to handle the additional (noise) features automatically (the user should not need to sample noise by hand).
> 
> Note that in the paper, Fig 2a represents a sequence of two coupling transforms and Fig 2c represents a sequence of two auto-regressive transforms. These are already implemented in Zuko.
> 
> If you wish to contribute to Zuko, maybe you should start with a feature that is easier to add. For example, the rational-linear spline transformation (see https://arxiv.org/abs/2001.05168) should be easy to implement. You can take inspiration from the rational-quadratic spline transformation, already present in the library:
> 
> https://github.com/probabilists/zuko/blob/6fd97cbd5d3a394be01a90ce094cb2343383a7f5/zuko/transforms.py

Thanks for your specific comment and suggestion! Because of my lack of experience, implementing ANF seems not-suitable at now. I will try rational-linear spline transformation first refering rational-quadratic spline transformation; Thanks for your comment again
